### PR TITLE
chore(auth)

### PR DESCRIPTION
**Pull request by kaustavha**
Fix/clean/rebase pr for ldap and modularizing the auth module more
This is a clone of #66 without their newrelic mods since a newer newrelic based mod exists on our master, cleaned up, reviewed, and configured to preserve backwards compatibility better. 

- Auth section is now abstracted out into separate auth handlers (htpasswd, passthru, ~~ldap~~)
- ~~Implentation of LDAP auth (using ldapauth npm module)~~
- Configuration (local_settings) can be used to choose which auth_method to use

### CODE REVIEW

**Comment by mburns** [comment]
@kaustavha please rebase.

**Comment by mburns** [comment]
@kaustavha ping

**Comment by kaustavha** [comment]
@mburns pong

File: lib/web/auth.js [hunk 1]
```diff
@@ -14,176 +14,28 @@
...
-    callback(null, false);
-  } else if (this.users[username].indexOf('{SHA}') === 0) {
-    callback(null, this._validateSHA(password, this.users[username]));
-  } else if (this.users[username].indexOf('$apr1$') === 0) {
-    callback(null, this._validateMD5(password, this.users[username]));
-  } else {
-    callback(null, false);
+// read all possible auth handlers (*.js from ./auth folder)
+fs.readdirSync(__dirname + '/auth').forEach(function(authMethod) {
+  if (authMethod.substr(-3) === '.js') {
```
Comment (rphillips) [inline lib/web/auth.js]: export a 'name' from the js file instead of relying on the filename.

File: lib/web/auth/htpasswd.js [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+    var db = new AuthDB(getUsers(filePath));
+
+    process.on('SIGHUP', function() {
+        db.updateUsers(getUsers(filePath));
+    });
+
+    return db;
+}
+
+exports.initialize = loadDBFromFile;
```
Comment (rphillips) [inline lib/web/auth/htpasswd.js]: loadDbFromFile should be initialize

File: lib/web/auth.js [hunk 3]
```diff
@@ -14,176 +14,28 @@
...
-  var rearranged = '',
-      mixin, final, m, v, i;
-
-  m = crypto.createHash('md5');
-  m.update(password + magic + salt);
-  mixin = crypto.createHash('md5').update(password + salt + password).digest("binary");
-
-  for (i = 0; i < password.length; i++) {
-    m.update(mixin[i % 16]);
+  if (Object.keys(authMethods).indexOf(config.auth_method) < 0) {
```
Comment (rphillips) [inline lib/web/auth.js]: Object.keys(authmethods) is called twice... it makes sense to pull them out into a variable.

File: lib/web/auth/passthru.js [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    } else {
+        callback(null, false);
+    }
+};
+
+function createProxyPassAuth() {
+    return new PassThruAuthDb();
+}
+
+exports.initialize = createProxyPassAuth;
```
Comment (rphillips) [inline lib/web/auth/passthru.js]: same here... should be initialize.

**Comment by kaustavha** [comment]
Right so:
- I forgot to add the ldap dep from the original pr in the package.json of this pr, when i added it the build started to fail on all versions of node above 0.10
- I support the refactor of the auth module, it cleans it up and makes it more modular
- It is currently out of scope to add ldap auth via this pr because the library it depends on is uh not good because:
  - It's old and dead, not touched since 2013
    https://github.com/trentm/node-ldapauth/commits/master
  - It relies on a git sha of another ldap library which imo is a little sketchy
    https://github.com/trentm/node-ldapauth/blob/master/package.json#L21
  - It relies on an ancient version of bcrypt which causes the build failures on node >v0.10
    https://github.com/trentm/node-ldapauth/blob/master/package.json#L20
    https://github.com/dcodeIO/bcrypt.js

I propose we scrap the ldap portion of this PR and merge in the auth module refactor, create an issue for ldap integration and revisit it in the future.

**Comment by mburns** [comment]
>I propose we scrap the ldap portion of this PR and merge in the auth module refactor, create an issue for ldap integration and revisit it in the future.

@kaustavha ok, let's do that then and wrap this up. I'd like to land these improvements.

**Comment by kaustavha** [comment]
That should do it.

**Comment by mburns** [comment]
(doesn't work in staging)

**Pull request closed without merging**